book-content/chapters/15-designing-your-types.md

```diff
@@ -0,0 +1,1444 @@
+  {
+    name: string;
+    tracks: string[];
+  },
+  {
+    creator: string;
+    artwork: string;
+    dateUpdated: Date;
+  }
+>;
+```
+
+Once the `id` property is added to the `TContent` type parameter, the error will go away.
+
+#### Constraints Describe Required Properties
+
+Note that these constraints we're providing here are just descriptions for properties the object must contain. We can pass `name` and `tracks` into `TContent` as long as it has an `id` property.
+
+In other words, these constraints are _open_, not _closed_. You won't get excess property warnings here. Any excess properties you pass in will be added to the type.
+
+#### `extends`, `extends`, `extends`
+
+By now, we've seen `extends` used in a few different contexts:
+
+- In generic types, to set constraints on type parameters
+- In classes, to extend another class
+- In interfaces, to extend another interface
+
+There is even another use for `extends` - conditional types, which we'll look at later in this chapter.
+
+One of TypeScript's annoying habits is that it tends to reuse the same keywords in different contexts. So it's important to understand that `extends` means different things in different places.
+
+## Template Literal Types in TypeScript
+
+Similar to how template literals in JavaScript allow you to interpolate values into strings, template literal types in TypeScript can be used to interpolate other types into string types.
+
+For example, let's create a `PngFile` type that represents a string that ends with ".png":
+
+```tsx
+type PngFile = `${string}.png`;
+```
+
+Now when we type a new variable as `PngFile`, it must end with ".png":
+
+```tsx
+let myImage: PngFile = "my-image.png"; // OK
+```
+
+When a string does not match the pattern defined in the `PngFile` type, TypeScript will raise an error:
+
+```ts twoslash
+// @errors: 2322
+type PngFile = `${string}.png`;
+
+// ---cut---
+let myImage: PngFile = "my-image.jpg";
+```
+
+Enforce specific string formats with template literal types can be useful in your applications.
+
+### Combining Template Literal Types with Union Types
+
+Template literal types become even more powerful when combined with union types. By passing a union to a template literal type, you can generate a type that represents all possible combinations of the union.
+
+For example, let's imagine we have a set of colors each with possible shades from `100` to `900`:
+
+```tsx
+type ColorShade = 100 | 200 | 300 | 400 | 500 | 600 | 700 | 800 | 900;
+type Color = "red" | "blue" | "green";
+```
+
+If we want a combination of all possible colors and shades, we can use a template literal type to generate a new type:
+
+```tsx
+type ColorPalette = `${Color}-${ColorShade}`;
+```
+
+Now, `ColorPalette` will represent all possible combinations of colors and shades:
+
+```tsx
+let myColor: ColorPalette = "red-500"; // OK
+let myColor2: ColorPalette = "blue-900"; // OK
+```
+
+That's 27 possible combinations - three colors times nine shades.
```

Comment:
This means that for big inputs u end up with even bigger outputs and u might hit union size limits